src/host_abi.zig

```diff
@@ -0,0 +1,108 @@
+    ops: *RocOps,
+    /// The argument that the Roc entrypoint function will receive from the host.
+    /// (If multiple arguments are needed, this should be a pointer to a struct.)
+    arg: *anyopaque,
+    /// What the Roc entrypoint function will return to the host.
+    /// (The caller should have allocated enough space for the Roc function to write
+    /// the entire return value into this.)
+    ret: *anyopaque,
+};
+
+/// The operations (in the form of function pointers) that a running Roc program
+/// needs the host to provide.
+///
+/// This is used in both calls from actual hosts as well as evaluation of constants
+/// inside the Roc compiler itself.
+pub const RocOps = struct {
+    /// The host provides this pointer, and Roc passes it to each of the following
+    /// function pointers as a second argument. This lets the host do things like use
+    /// arena allocators for allocation and deallocation (by putting the arena in here).
+    /// The pointer can be to absolutely anything the host likes, or null if unused.
+    env: *anyopaque,
+    /// Like _aligned_malloc (size, alignment) - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-malloc
+    /// Roc will automatically call roc_crashed if this returns null.
+    roc_alloc: fn (*RocAlloc, *anyopaque) void,
+    /// Like _aligned_free - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-free
+    roc_dealloc: fn (*RocDealloc, *anyopaque) void,
+    /// Like _aligned_realloc (ptr, size, alignment) - https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/aligned-realloc
+    /// Roc will automatically call roc_crashed if this returns null.
+    roc_realloc: fn (*RocRealloc, *anyopaque) void,
+    /// Called when the Roc program crashes, e.g. due to integer overflow.
+    /// It receives a pointer to a UTF-8 string, along with its length in bytes.
+    /// This function must not return, because the Roc program assumes it will
+    /// not continue to be executed after this function is called.
+    roc_crashed: fn (*RocCrashed, *anyopaque) void,
+    /// Called when the Roc program has called `dbg` on something.
+    roc_dbg: fn (*RocDbg, *anyopaque) void,
+    /// Called when the Roc program has run an `expect` which failed.
+    roc_expect_failed: fn (*RocExpectFailed, *anyopaque) void,
+};
+
+/// When RocOps.roc_alloc gets called, it will be passed one of these.
+/// That function should write the allocated memory into `ret`.
+/// If it cannot proivde a non-null pointer (e.g. due to OOM), it
+/// must not return, and must instead do something along the lines
+/// of roc_crashed.
+pub const RocAlloc = struct {
+    alignment: usize,
+    length: usize,
+    ret: *anyopaque,
+};
+
+/// When RocOps.roc_dealloc gets called, it will be passed one of these.
+/// (The length of the allocation cannot be provided, because it is
+/// not always known at runtime due to the way seamless slices work.)
```

Comment:
This is something I really wish we could fix, but I don't think it is feasible.

Reply:
Agreed on both points!